### ISSUE TITLE
NoSuchMethodError when running a job on DSE 4.6 with Cassandra Connector

### ISSUE THREAD

**Issue by cloudify** [spark-jobserver/spark-jobserver#114]
Hi there,

I'm opening an issue because I'm running out of options! :smile:

I have a Datastax Enterprise cluster (v4.6) running Spark 1.1, I installed the job server (0.4.1) on the master node and everything works fine for spark jobs that don't use the cassandra connector.

Problem is that I'm unable to run any job that make use of the cassandra connector, the result is the following exception in the jobserver log, following the shutdown of the jobserver:

```
java.lang.NoSuchMethodError: org.apache.thrift.EncodingUtils.setBit(BIZ)B
	at org.apache.cassandra.thrift.Cassandra$describe_splits_ex_args.setKeys_per_splitIsSet(Cassandra.java:39993)
	at org.apache.cassandra.thrift.Cassandra$describe_splits_ex_args.setKeys_per_split(Cassandra.java:39979)
	at org.apache.cassandra.thrift.Cassandra$Client.send_describe_splits_ex(Cassandra.java:1448)
	at org.apache.cassandra.thrift.Cassandra$Client.describe_splits_ex(Cassandra.java:1438)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
...
```

A couple of Google searches tell me that this error usually happens when there's multiple thrift libraries in the classpath, and in fact DSE installs two different versions of `libthrift` and they're both in the classpath:

```
$ find . -type f | grep libthrift
./shark/lib/libthrift-0.9.1.jar
./spark/lib/libthrift-0.8.0.jar
./hive/lib/libthrift-0.9.1.jar
./cassandra/lib/libthrift-0.9.1.jar
```

The job I'm running is very simple, it's something like:

```
  override def runJob(sc: SparkContext, config: Config): Any = {
    val rdd = sc.cassandraTable("mykeyspace", "mytable")
    rdd.count()
  }
```

library dependencies:

```
  "spark.jobserver"          % "job-server-api"             % "0.4.1" % "provided",
  "org.apache.spark"        %% "spark-core"                 % "1.1.0" % "provided",
  "com.datastax.spark"      %% "spark-cassandra-connector"  % "1.1.1" % "provided"
```

I tried the following:

* remove the `provided` flag on `spark-cassandra-connector` (thus embedding the connector in the fat jar) => exception
* embed `libthrift-0.9.1.jar` in the fat jat => exception
* run the same code with `spark-submit` => works fine!!! (I've analyzed the classpaths used to run the job by the workers for both the jobserver and the `spark-submit` version and they match, they both contain `libthrift-0.8.0.jar` and `libthrift-0.9.1.jar` in the same order)
* adding `libthrift-0.9.1` to the head of the classpath with the `dependent-jar-uris` config (confirmed by looking at the cp in the worker logs) => exception
* tried the same with `spark.executor.extraClassPath` => exception

Any help is appreciated! :smile: 

thanks

**Comment by cloudify** [comment]
So I verified that in fact when the job is run by the jobserver, `libthrift-0.8.0.jar` gets loaded, but the run by spark-submit `libthrift-0.9.1.jar` gets loaded. So I'm assuming the classes are loaded in different ways.

I found a quick "fix" by removing `libthrift-0.8.0.jar` from the classpath generated by spark and passed to the jobserver, something like:

```
CLASSPATH="$appdir:$appdir/spark-job-server.jar:$($SPARK_HOME/bin/compute-classpath.sh|sed 's/:\/usr\/share\/dse\/spark\/lib\/libthrift-0.8.0.jar//')"
```

Everything works fine now, but I'd love to know if there's a more elegant solution.

thanks

**Comment by velvia** [comment]
Hi Federico,

Hmmm, yeah that's definitely not an elegant solution, glad you found one
though and are sharing it.

I assume you are using the start script provided by job server?  There must
be differences in the scripts....  perhpas you could start job server using
spark-submit, which would in theory work too.

-Evan

On Sat, Apr 4, 2015 at 5:00 AM, Federico Feroldi <notifications@github.com>
wrote:

> So I verified that in fact when the job is run by the jobserver,
> libthrift-0.8.0.jar gets loaded, but the run by spark-submit
> libthrift-0.9.1.jar gets loaded. So I'm assuming the classes are loaded
> in different ways.
>
> I found a quick "fix" by removing libthrift-0.8.0.jar from the classpath
> generated by spark and passed to the jobserver, something like:
>
> CLASSPATH="$appdir:$appdir/spark-job-server.jar:$($SPARK_HOME/bin/compute-classpath.sh|sed 's/:\/usr\/share\/dse\/spark\/lib\/libthrift-0.8.0.jar//')"
>
> Everything works fine now, but I'd love to know if there's a more elegant
> solution.
>
> thanks
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/issues/114#issuecomment-89556754>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

**Comment by velvia** [comment]
@cloudify have you tried DSE 4.8 that includes the job server?   Will close this for now.
